src/win/tty.c

```diff
@@ -937,30 +956,106 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
+    /* Cancel line-buffered read if not already pending */
+    err = uv__cancel_read_console(handle);
+    if (err)
+      return err;
+
+    handle->flags |= UV_HANDLE_CANCELLATION_PENDING;
+  }
+
+  return 0;
+}
+
+static int uv__cancel_read_console(uv_tty_t* handle) {
+  HANDLE active_screen_buffer = INVALID_HANDLE_VALUE;
+  CONSOLE_SCREEN_BUFFER_INFO info;
+  BOOL restore_screen_state;
+  INPUT_RECORD record;
+  DWORD written;
+  DWORD err = 0;
+
+  EnterCriticalSection(&uv__read_console_trap_lock);
+  if (!uv__read_console_started) {
+    /* We have managed to trap the other thread before ReadConsole is called */
+    uv__read_console_trap = TRUE;
+    LeaveCriticalSection(&uv__read_console_trap_lock);
+    return 0;
   }
+  LeaveCriticalSection(&uv__read_console_trap_lock);
+
+  /* If ReadConsole already returned (the user pressed ENTER),
+     there is nothing to do */
+  if (WaitForSingleObject(uv__read_console_done_event, 0) == WAIT_OBJECT_0)
+    return 0;
+
+  /* From this point return should only happen by jumping to the out label
+     to ensure proper cleanup */
+
+  /* Save screen state before sending the VK_RETURN event */
+  active_screen_buffer = CreateFileA("conout$",
+                                     GENERIC_READ | GENERIC_WRITE,
+                                     FILE_SHARE_READ | FILE_SHARE_WRITE,
+                                     NULL,
+                                     OPEN_EXISTING,
+                                     FILE_ATTRIBUTE_NORMAL,
+                                     NULL);
 
-  /* Cancel line-buffered read */
-  if (handle->tty.rd.read_line_handle != NULL) {
-    /* Closing this handle will cancel the ReadConsole operation */
-    CloseHandle(handle->tty.rd.read_line_handle);
-    handle->tty.rd.read_line_handle = NULL;
+  if (active_screen_buffer != INVALID_HANDLE_VALUE &&
+      GetConsoleScreenBufferInfo(active_screen_buffer, &info)) {
+    restore_screen_state = TRUE;
+  } else {
+    restore_screen_state = FALSE;
+  }
+
+  /* Write enter key event to force the console wait to return. */
+  record.EventType = KEY_EVENT;
+  record.Event.KeyEvent.bKeyDown = TRUE;
+  record.Event.KeyEvent.wRepeatCount = 1;
+  record.Event.KeyEvent.wVirtualKeyCode = VK_RETURN;
+  record.Event.KeyEvent.wVirtualScanCode =
+    MapVirtualKeyW(VK_RETURN, MAPVK_VK_TO_VSC);
+  record.Event.KeyEvent.uChar.UnicodeChar = L'\r';
+  record.Event.KeyEvent.dwControlKeyState = 0;
+  if (!WriteConsoleInputW(handle->handle, &record, 1, &written)) {
+    err = GetLastError();
+    goto out;
+  } else {
+    /* At this point, we have sent the VK_RETURN.
+       Any further errors shouldn't be be returned. */
+    err = 0;
   }
 
+  WaitForSingleObject(uv__read_console_done_event, INFINITE);
 
-  return 0;
+  /* Restore the screen state to undo the effect of the VK_RETURN we sent */
+  if (!restore_screen_state)
+    goto out;
+
+  SetConsoleCursorPosition(active_screen_buffer, info.dwCursorPosition);
```

Comment:
Yes, this is a corner case that I know could be improved. At first I implemented this reusing the existing uv_tty_save_state/uv_tty_restore_state logic, but that had some apparent problems. After switching to the simpler approach, I haven't been able to produce any problems with scrolling in any of my tests, though I admit I was under time pressure to deliver this fix. Perhaps we can fix it in a second iteration? It should be a very rare case, and even a cursor jump is a minor issue compared to the current issue.